subt/docker/robotika/ros_proxy_node.cc

```diff
@@ -0,0 +1,531 @@
+#include <rosgraph_msgs/Clock.h>
+#include <sensor_msgs/Imu.h>
+#include <sensor_msgs/LaserScan.h>
+#include <sensor_msgs/CompressedImage.h>
+#include <nav_msgs/Odometry.h>
+
+#include <subt_msgs/PoseFromArtifact.h>
+#include <ros/ros.h>
+#include <std_srvs/SetBool.h>
+#include <rosgraph_msgs/Clock.h>
+
+#include <string>
+#include <stdlib.h>     /* abs */
+
+#include <subt_communication_broker/subt_communication_client.h>
+#include <subt_ign/CommonTypes.hh>
+#include <subt_ign/protobuf/artifact.pb.h>
+
+// PROXY PART
+#include <zmq.h>
+#include <assert.h>
+
+int g_countClock = 0;
+int g_countImu = 0;
+int g_countScan = 0;
+int g_countImage = 0;
+int g_countOdom = 0;
+
+void *g_context;
+void *g_responder;
+
+void *g_contextIn;
+void *g_requester;
+
+void initZeroMQ()
+{
+  g_context = zmq_ctx_new ();
+  g_responder = zmq_socket (g_context, ZMQ_PUSH);  // use "Pipeline pattern" to send all data to Python3
+  int rc = zmq_bind (g_responder, "tcp://*:5555");
+  assert (rc == 0);
+
+  // received
+  g_contextIn = zmq_ctx_new ();
+  g_requester = zmq_socket (g_contextIn, ZMQ_PULL);  // use "Pipeline pattern" to receive all data to Python3
+  rc = zmq_bind (g_requester, "tcp://*:5556");
+  assert (rc == 0);
+}
+
+void clockCallback(const rosgraph_msgs::Clock::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  zmq_send(g_responder, sm.buf.get(), sm.num_bytes, 0);
+  if(g_countClock % 1000 == 0)
+    ROS_INFO("received Clock %d ", g_countClock);
+  g_countClock++;
+}
+
+void imuCallback(const sensor_msgs::Imu::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  zmq_send(g_responder, sm.buf.get(), sm.num_bytes, 0);
+  if(g_countImu % 100 == 0)
+    ROS_INFO("received Imu %d ", g_countImu);
+  g_countImu++;
+}
+
+void scanCallback(const sensor_msgs::LaserScan::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  zmq_send(g_responder, sm.buf.get(), sm.num_bytes, 0);
+  if(g_countScan % 100 == 0)
+    ROS_INFO("received Scan %d", g_countScan);
+  g_countScan++;
+}
+
+void imageCallback(const sensor_msgs::CompressedImage::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  zmq_send(g_responder, sm.buf.get(), sm.num_bytes, 0);
+  if(g_countImage % 100 == 0)
+    ROS_INFO("received Image %d", g_countImage);
+  g_countImage++;
+}
+
+void odomCallback(const nav_msgs::Odometry::ConstPtr& msg)
+{
+  ros::SerializedMessage sm = ros::serialization::serializeMessage(*msg);
+  zmq_send(g_responder, sm.buf.get(), sm.num_bytes, 0);
+  if(g_countOdom % 100 == 0)
+    ROS_INFO("received Odom %d", g_countOdom);
+  g_countOdom++;
+}
+
+void sendOrigin(std::string& name, double x, double y, double z,
+                double qx, double qy, double qz, double qw)
+{
+  char buf[1000];
+  int size = sprintf(buf, "origin %s %lf %lf %lf  %lf %lf %lf %lf", name.c_str(), x, y, z, qx, qy, qz, qw);
```

Comment:
The idiomatic way to do this in C++ is to use stringstream. See for example https://stackoverflow.com/a/20595061/3185929.